File: mRemoteV1/Config/Config.Settings.Load.vb
Imports System.Globalization
Imports System.IO
Imports System.Threading
Imports System.Xml
Imports mRemote3G.App
Imports mRemote3G.App.Info
Imports mRemote3G.Connection.Protocol
Imports mRemote3G.Forms
Imports mRemote3G.My
Imports mRemote3G.Security
Imports mRemote3G.Themes
Imports mRemote3G.Tools
Imports mRemote3G.UI.Window
Imports WeifenLuo.WinFormsUI.Docking

Namespace Config

    Namespace Settings
        Public Class Load

#Region "Public Properties"

            Private _MainForm As frmMain

            Public Property MainForm As frmMain
                Get
                    Return Me._MainForm
                End Get
                Set
                    Me._MainForm = value
                End Set
            End Property

#End Region

#Region "Public Methods"

            Public Sub New(MainForm As frmMain)
                Me._MainForm = MainForm
            End Sub

            Public Sub Load()
                Try
                    With Me._MainForm
                        ' Migrate settings from previous version
                        If MySettingsProperty.Settings.DoUpgrade Then
                            Try
                                MySettingsProperty.Settings.Upgrade()
                            Catch ex As Exception
                                Runtime.Log.Error("My.Settings.Upgrade() failed" & vbNewLine & ex.ToString())
                            End Try
                            MySettingsProperty.Settings.DoUpgrade = False

                            ' Clear pending update flag
                            ' This is used for automatic updates, not for settings migration, but it
                            ' needs to be cleared here because we know that we just updated.
                            MySettingsProperty.Settings.UpdatePending = False
                        End If

                        SupportedCultures.InstantiateSingleton()
                        If _
                            Not MySettingsProperty.Settings.OverrideUICulture = "" And
                            SupportedCultures.IsNameSupported(MySettingsProperty.Settings.OverrideUICulture) Then
                            Thread.CurrentThread.CurrentUICulture =
                                New CultureInfo(MySettingsProperty.Settings.OverrideUICulture)
                            Runtime.Log.InfoFormat("Override Culture: {0}/{1}",
                                                   Thread.CurrentThread.CurrentUICulture.Name,
                                                   Thread.CurrentThread.CurrentUICulture.NativeName)
                        End If

                        ThemeManager.LoadTheme(MySettingsProperty.Settings.ThemeName)

                        .WindowState = FormWindowState.Normal
                        If MySettingsProperty.Settings.MainFormState = FormWindowState.Normal Then
                            If Not MySettingsProperty.Settings.MainFormLocation.IsEmpty Then _
                                .Location = MySettingsProperty.Settings.MainFormLocation
                            If Not MySettingsProperty.Settings.MainFormSize.IsEmpty Then _
                                .Size = MySettingsProperty.Settings.MainFormSize
                        Else
                            If Not MySettingsProperty.Settings.MainFormRestoreLocation.IsEmpty Then _
                                .Location = MySettingsProperty.Settings.MainFormRestoreLocation
                            If Not MySettingsProperty.Settings.MainFormRestoreSize.IsEmpty Then _
                                .Size = MySettingsProperty.Settings.MainFormRestoreSize
                        End If
                        If MySettingsProperty.Settings.MainFormState = FormWindowState.Maximized Then
                            .WindowState = FormWindowState.Maximized
                        End If

                        ' Make sure the form is visible on the screen
                        Const minHorizontal = 300
                        Const minVertical = 150
                        Dim screenBounds As Rectangle = Screen.FromHandle(.Handle).Bounds
                        Dim newBounds As Rectangle = .Bounds

                        If newBounds.Right < screenBounds.Left + minHorizontal Then
                            newBounds.X = screenBounds.Left + minHorizontal - newBounds.Width
                        End If
                        If newBounds.Left > screenBounds.Right - minHorizontal Then
                            newBounds.X = screenBounds.Right - minHorizontal
                        End If
                        If newBounds.Bottom < screenBounds.Top + minVertical Then
                            newBounds.Y = screenBounds.Top + minVertical - newBounds.Height
                        End If
                        If newBounds.Top > screenBounds.Bottom - minVertical Then
                            newBounds.Y = screenBounds.Bottom - minVertical
                        End If

                        .Location = newBounds.Location

                        If MySettingsProperty.Settings.MainFormKiosk = True Then
                            .Fullscreen.Value = True
                            .mMenViewFullscreen.Checked = True
                        End If

                        If MySettingsProperty.Settings.UseCustomPuttyPath Then
                            PuttyBase.PuttyPath = MySettingsProperty.Settings.CustomPuttyPath
                        Else
                            PuttyBase.PuttyPath = General.PuttyPath
                        End If

                        If MySettingsProperty.Settings.ShowSystemTrayIcon Then
                            Runtime.NotificationAreaIcon = New Tools.Controls.NotificationAreaIcon()
                        End If

                        If MySettingsProperty.Settings.AutoSaveEveryMinutes > 0 Then
                            .tmrAutoSave.Interval = MySettingsProperty.Settings.AutoSaveEveryMinutes*60000
                            .tmrAutoSave.Enabled = True
                        End If

                        MySettingsProperty.Settings.ConDefaultPassword =
                            Crypt.Decrypt(MySettingsProperty.Settings.ConDefaultPassword, General.EncryptionKey)

                        Me.LoadPanelsFromXML()
                        Me.LoadExternalAppsFromXML()

                        If MySettingsProperty.Settings.AlwaysShowPanelTabs Then
                            frmMain.pnlDock.DocumentStyle = DocumentStyle.DockingWindow
                        End If

                        If MySettingsProperty.Settings.ResetToolbars = False Then
                            LoadToolbarsFromSettings()
                        Else
                            SetToolbarsDefault()
                        End If
                    End With
                Catch ex As Exception
                    Runtime.Log.Error("Loading settings failed" & vbNewLine & ex.ToString())
                End Try
            End Sub

            Public Sub SetToolbarsDefault()
                With MainForm
                    ToolStripPanelFromString("top").Join(.tsQuickConnect, New Point(300, 0))
                    .tsQuickConnect.Visible = True
                    ToolStripPanelFromString("bottom").Join(.tsExternalTools, New Point(3, 0))
                    .tsExternalTools.Visible = False
                End With
            End Sub

            Public Sub LoadToolbarsFromSettings()
                With Me.MainForm
                    If MySettingsProperty.Settings.QuickyTBLocation.X > MySettingsProperty.Settings.ExtAppsTBLocation.X _
                        Then
                        AddDynamicPanels()
                        AddStaticPanels()
                    Else
                        AddStaticPanels()
                        AddDynamicPanels()
                    End If
                End With
            End Sub

            Private Sub AddStaticPanels()
                With MainForm
                    ToolStripPanelFromString(MySettingsProperty.Settings.QuickyTBParentDock).Join(.tsQuickConnect, MySettingsProperty.Settings.QuickyTBLocation)
                    .tsQuickConnect.Visible = MySettingsProperty.Settings.QuickyTBVisible
                End With
            End Sub

            Private Sub AddDynamicPanels()
                With MainForm
                    ToolStripPanelFromString(MySettingsProperty.Settings.ExtAppsTBParentDock).Join(.tsExternalTools, MySettingsProperty.Settings.ExtAppsTBLocation)
                    .tsExternalTools.Visible = MySettingsProperty.Settings.ExtAppsTBVisible
                End With
            End Sub

            Private Function ToolStripPanelFromString(Panel As String) As ToolStripPanel
                Select Case LCase(Panel)
                    Case "top"
                        Return MainForm.tsContainer.TopToolStripPanel
                    Case "bottom"
                        Return MainForm.tsContainer.BottomToolStripPanel
                    Case "left"
                        Return MainForm.tsContainer.LeftToolStripPanel
                    Case "right"
                        Return MainForm.tsContainer.RightToolStripPanel
                    Case Else
                        Return MainForm.tsContainer.TopToolStripPanel
                End Select
            End Function

            Public Sub LoadPanelsFromXML()
                Try
                    With MainForm
                        Runtime.Windows.treePanel = Nothing
                        Runtime.Windows.configPanel = Nothing
                        Runtime.Windows.errorsPanel = Nothing

                        Do While .pnlDock.Contents.Count > 0
                            Dim dc As DockContent = .pnlDock.Contents(0)
                            dc.Close()
                        Loop

                        Runtime.Startup.CreatePanels()

                        Dim oldPath As String =
                                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) & "\" &
                                Application.Info.ProductName & "\" & Info.Settings.LayoutFileName
                        Dim newPath As String = Info.Settings.SettingsPath & "\" & Info.Settings.LayoutFileName
                        If File.Exists(newPath) Then
                            .pnlDock.LoadFromXml(newPath, AddressOf GetContentFromPersistString)
#If Not PORTABLE Then
                        ElseIf File.Exists(oldPath) Then
                            .pnlDock.LoadFromXml(oldPath, AddressOf GetContentFromPersistString)
#End If
                        Else
                            Runtime.Startup.SetDefaultLayout()
                        End If
                    End With
                Catch ex As Exception
                    Runtime.Log.Error("LoadPanelsFromXML failed" & vbNewLine & ex.ToString())
                End Try
            End Sub

            Public Sub LoadExternalAppsFromXML()
                Dim oldPath As String = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) & "\" &
                                        Application.Info.ProductName & "\" & Info.Settings.ExtAppsFilesName
                Dim newPath As String = Info.Settings.SettingsPath & "\" & Info.Settings.ExtAppsFilesName
                Dim xDom As New XmlDocument()
                If File.Exists(newPath) Then
                    xDom.Load(newPath)
#If Not PORTABLE Then
                ElseIf File.Exists(oldPath) Then
                    xDom.Load(oldPath)
#End If
                Else
                    Exit Sub
                End If

                For Each xEl As XmlElement In xDom.DocumentElement.ChildNodes
                    Dim extA As New ExternalTool
                    extA.DisplayName = xEl.Attributes("DisplayName").Value
                    extA.FileName = xEl.Attributes("FileName").Value
                    extA.Arguments = xEl.Attributes("Arguments").Value

                    If xEl.HasAttribute("WaitForExit") Then
                        extA.WaitForExit = xEl.Attributes("WaitForExit").Value
                    End If

                    If xEl.HasAttribute("TryToIntegrate") Then
                        extA.TryIntegrate = xEl.Attributes("TryToIntegrate").Value
                    End If

                    Runtime.ExternalTools.Add(extA)
                Next

                MainForm.SwitchToolBarText(MySettingsProperty.Settings.ExtAppsTBShowText)

                xDom = Nothing

                frmMain.AddExternalToolsToToolBar()
            End Sub

#End Region

#Region "Private Methods"

            Private Function GetContentFromPersistString(persistString As String) As IDockContent
                ' pnlLayout.xml persistence XML fix for refactoring to mRemoteNG
                If (persistString.StartsWith("mRemote.")) Then
                    persistString = persistString.Replace("mRemote.", "mRemote3G.")
                End If

                Try
                    If persistString = GetType(UI.Window.Config).ToString Then
                        Return Runtime.Windows.configPanel
                    End If

                    If persistString = GetType(UI.Window.Tree).ToString Then
                        Return Runtime.Windows.treePanel
                    End If

                    If persistString = GetType(ErrorsAndInfos).ToString Then
                        Return Runtime.Windows.errorsPanel
                    End If

                    If persistString = GetType(ScreenshotManager).ToString Then
                        Return Runtime.Windows.screenshotPanel
                    End If
                Catch ex As Exception
                    Runtime.Log.Error("GetContentFromPersistString failed" & vbNewLine & ex.ToString())
                End Try

                Return Nothing
            End Function

#End Region
        End Class
    End Namespace

End Namespace

File: mRemoteV1/CS/Config/Config.Settings.Load.cs
using System;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Windows.Forms;
using System.IO;
using WeifenLuo.WinFormsUI.Docking;
using mRemoteNG.App;
using System.Xml;


namespace mRemoteNG.Config.Settings
{
	public class Load
	{
        #region Public Properties
		private frmMain _MainForm;
        public frmMain MainForm
		{
			get
			{
				return this._MainForm;
			}
			set
			{
				this._MainForm = value;
			}
		}
        #endregion
				
        #region Public Methods
		public Load(frmMain MainForm)
		{
			this._MainForm = MainForm;
		}
				
		public void LoadSettings()
		{
			try
			{
				// Migrate settings from previous version
				if (My.Settings.Default.DoUpgrade)
				{
					try
					{
						My.Settings.Default.Upgrade();
					}
					catch (Exception ex)
					{
                        Runtime.Log.Error("My.Settings.Upgrade() failed" + Constants.vbNewLine + ex.Message);
					}
					My.Settings.Default.DoUpgrade = false;
							
					// Clear pending update flag
					// This is used for automatic updates, not for settings migration, but it
					// needs to be cleared here because we know that we just updated.
					My.Settings.Default.UpdatePending = false;
				}
						
				App.SupportedCultures.InstantiateSingleton();
				if (!(My.Settings.Default.OverrideUICulture == "") && App.SupportedCultures.IsNameSupported(System.Convert.ToString(My.Settings.Default.OverrideUICulture)))
				{
					System.Threading.Thread.CurrentThread.CurrentUICulture = new System.Globalization.CultureInfo(System.Convert.ToString(My.Settings.Default.OverrideUICulture));
					Runtime.Log.InfoFormat("Override Culture: {0}/{1}", System.Threading.Thread.CurrentThread.CurrentUICulture.Name, System.Threading.Thread.CurrentThread.CurrentUICulture.NativeName);
				}
						
				Themes.ThemeManager.LoadTheme(System.Convert.ToString(My.Settings.Default.ThemeName));
						
				this._MainForm.WindowState = FormWindowState.Normal;
				if (My.Settings.Default.MainFormState == FormWindowState.Normal)
				{
					if (!My.Settings.Default.MainFormLocation.IsEmpty)
					{
						this._MainForm.Location = My.Settings.Default.MainFormLocation;
					}
					if (!My.Settings.Default.MainFormSize.IsEmpty)
					{
						this._MainForm.Size = My.Settings.Default.MainFormSize;
					}
				}
				else
				{
					if (!My.Settings.Default.MainFormRestoreLocation.IsEmpty)
					{
						this._MainForm.Location = My.Settings.Default.MainFormRestoreLocation;
					}
					if (!My.Settings.Default.MainFormRestoreSize.IsEmpty)
					{
						this._MainForm.Size = My.Settings.Default.MainFormRestoreSize;
					}
				}
				if (My.Settings.Default.MainFormState == FormWindowState.Maximized)
				{
					this._MainForm.WindowState = FormWindowState.Maximized;
				}
						
				// Make sure the form is visible on the screen
				const int minHorizontal = 300;
				const int minVertical = 150;
				System.Drawing.Rectangle screenBounds = Screen.FromHandle(this._MainForm.Handle).Bounds;
				System.Drawing.Rectangle newBounds = this._MainForm.Bounds;
						
				if (newBounds.Right < screenBounds.Left + minHorizontal)
				{
					newBounds.X = screenBounds.Left + minHorizontal - newBounds.Width;
				}
				if (newBounds.Left > screenBounds.Right - minHorizontal)
				{
					newBounds.X = screenBounds.Right - minHorizontal;
				}
				if (newBounds.Bottom < screenBounds.Top + minVertical)
				{
					newBounds.Y = screenBounds.Top + minVertical - newBounds.Height;
				}
				if (newBounds.Top > screenBounds.Bottom - minVertical)
				{
					newBounds.Y = screenBounds.Bottom - minVertical;
				}
						
				this._MainForm.Location = newBounds.Location;
						
				if (My.Settings.Default.MainFormKiosk == true)
				{
					this._MainForm.Fullscreen.Value = true;
					this._MainForm.mMenViewFullscreen.Checked = true;
				}
						
				if (My.Settings.Default.UseCustomPuttyPath)
				{
					Connection.Protocol.PuttyBase.PuttyPath = System.Convert.ToString(My.Settings.Default.CustomPuttyPath);
				}
				else
				{
					Connection.Protocol.PuttyBase.PuttyPath = App.Info.General.PuttyPath;
				}
						
				if (My.Settings.Default.ShowSystemTrayIcon)
				{
					App.Runtime.NotificationAreaIcon = new Tools.Controls.NotificationAreaIcon();
				}
						
				if (My.Settings.Default.AutoSaveEveryMinutes > 0)
				{
					this._MainForm.tmrAutoSave.Interval = System.Convert.ToInt32(My.Settings.Default.AutoSaveEveryMinutes * 60000);
					this._MainForm.tmrAutoSave.Enabled = true;
				}
						
				My.Settings.Default.ConDefaultPassword = Security.Crypt.Decrypt(System.Convert.ToString(My.Settings.Default.ConDefaultPassword), App.Info.General.EncryptionKey);
						
				this.LoadPanelsFromXML();
				this.LoadExternalAppsFromXML();
						
				if (My.Settings.Default.AlwaysShowPanelTabs)
				{
					frmMain.Default.pnlDock.DocumentStyle = DocumentStyle.DockingWindow;
				}
						
				if (My.Settings.Default.ResetToolbars == false)
				{
					LoadToolbarsFromSettings();
				}
				else
				{
					SetToolbarsDefault();
				}
			}
			catch (Exception ex)
			{
                Runtime.Log.Error("Loading settings failed" + Constants.vbNewLine + ex.Message);
			}
		}
				
		public void SetToolbarsDefault()
		{
			ToolStripPanelFromString("top").Join(MainForm.tsQuickConnect, new Point(300, 0));
			MainForm.tsQuickConnect.Visible = true;
			ToolStripPanelFromString("bottom").Join(MainForm.tsExternalTools, new Point(3, 0));
			MainForm.tsExternalTools.Visible = false;
		}
				
		public void LoadToolbarsFromSettings()
		{
			if (My.Settings.Default.QuickyTBLocation.X > My.Settings.Default.ExtAppsTBLocation.X)
			{
				AddDynamicPanels();
				AddStaticPanels();
			}
			else
			{
				AddStaticPanels();
				AddDynamicPanels();
			}
		}
				
		private void AddStaticPanels()
		{
			ToolStripPanelFromString(System.Convert.ToString(My.Settings.Default.QuickyTBParentDock)).Join(MainForm.tsQuickConnect, My.Settings.Default.QuickyTBLocation);
			MainForm.tsQuickConnect.Visible = System.Convert.ToBoolean(My.Settings.Default.QuickyTBVisible);
		}
				
		private void AddDynamicPanels()
		{
			ToolStripPanelFromString(System.Convert.ToString(My.Settings.Default.ExtAppsTBParentDock)).Join(MainForm.tsExternalTools, My.Settings.Default.ExtAppsTBLocation);
			MainForm.tsExternalTools.Visible = System.Convert.ToBoolean(My.Settings.Default.ExtAppsTBVisible);
		}
				
		private ToolStripPanel ToolStripPanelFromString(string Panel)
		{
			switch (Panel.ToLower())
			{
				case "top":
					return MainForm.tsContainer.TopToolStripPanel;
				case "bottom":
					return MainForm.tsContainer.BottomToolStripPanel;
				case "left":
					return MainForm.tsContainer.LeftToolStripPanel;
				case "right":
					return MainForm.tsContainer.RightToolStripPanel;
				default:
					return MainForm.tsContainer.TopToolStripPanel;
			}
		}
				
		public void LoadPanelsFromXML()
		{
			try
			{
                Runtime.Windows.treePanel = null;
                Runtime.Windows.configPanel = null;
                Runtime.Windows.errorsPanel = null;
						
				while (MainForm.pnlDock.Contents.Count > 0)
				{
                    WeifenLuo.WinFormsUI.Docking.DockContent dc = (WeifenLuo.WinFormsUI.Docking.DockContent)MainForm.pnlDock.Contents[0];
					dc.Close();
				}

                Runtime.Startup.CreatePanels();
						
				string oldPath = System.Environment.GetFolderPath(System.Environment.SpecialFolder.LocalApplicationData) + "\\" + (new Microsoft.VisualBasic.ApplicationServices.WindowsFormsApplicationBase()).Info.ProductName + "\\" + App.Info.Settings.LayoutFileName;
				string newPath = App.Info.Settings.SettingsPath + "\\" + App.Info.Settings.LayoutFileName;
				if (File.Exists(newPath))
				{
					MainForm.pnlDock.LoadFromXml(newPath, GetContentFromPersistString);
                #if !PORTABLE
				}
				else if (File.Exists(oldPath))
				{
					MainForm.pnlDock.LoadFromXml(oldPath, GetContentFromPersistString);
                #endif
				}
				else
				{
                    Runtime.Startup.SetDefaultLayout();
				}
			}
			catch (Exception ex)
			{
                Runtime.Log.Error("LoadPanelsFromXML failed" + Constants.vbNewLine + ex.Message);
			}
		}
				
		public void LoadExternalAppsFromXML()
		{
			string oldPath = System.Environment.GetFolderPath(System.Environment.SpecialFolder.LocalApplicationData) + "\\" + (new Microsoft.VisualBasic.ApplicationServices.WindowsFormsApplicationBase()).Info.ProductName + "\\" + App.Info.Settings.ExtAppsFilesName;
			string newPath = App.Info.Settings.SettingsPath + "\\" + App.Info.Settings.ExtAppsFilesName;
			XmlDocument xDom = new XmlDocument();
			if (File.Exists(newPath))
			{
				xDom.Load(newPath);
            #if !PORTABLE
			}
			else if (File.Exists(oldPath))
			{
				xDom.Load(oldPath);
            #endif
			}
			else
			{
				return;
			}
					
			foreach (XmlElement xEl in xDom.DocumentElement.ChildNodes)
			{
				Tools.ExternalTool extA = new Tools.ExternalTool();
				extA.DisplayName = xEl.Attributes["DisplayName"].Value;
				extA.FileName = xEl.Attributes["FileName"].Value;
				extA.Arguments = xEl.Attributes["Arguments"].Value;
						
				if (xEl.HasAttribute("WaitForExit"))
				{
					extA.WaitForExit = bool.Parse(xEl.Attributes["WaitForExit"].Value);
				}
						
				if (xEl.HasAttribute("TryToIntegrate"))
				{
					extA.TryIntegrate = bool.Parse(xEl.Attributes["TryToIntegrate"].Value);
				}

                Runtime.ExternalTools.Add(extA);
			}
					
			MainForm.SwitchToolBarText(System.Convert.ToBoolean(My.Settings.Default.ExtAppsTBShowText));
					
			xDom = null;
					
			frmMain.Default.AddExternalToolsToToolBar();
		}
        #endregion
				
        #region Private Methods
		private IDockContent GetContentFromPersistString(string persistString)
		{
			// pnlLayout.xml persistence XML fix for refactoring to mRemoteNG
			if (persistString.StartsWith("mRemote."))
			{
				persistString = persistString.Replace("mRemote.", "mRemoteNG.");
			}
					
			try
			{
				if (persistString == typeof(UI.Window.Config).ToString())
				{
                    return Runtime.Windows.configPanel;
				}
						
				if (persistString == typeof(UI.Window.Tree).ToString())
				{
                    return Runtime.Windows.treePanel;
				}
						
				if (persistString == typeof(UI.Window.ErrorsAndInfos).ToString())
				{
                    return Runtime.Windows.errorsPanel;
				}
						
				if (persistString == typeof(UI.Window.Sessions).ToString())
				{
                    return Runtime.Windows.sessionsPanel;
				}
						
				if (persistString == typeof(UI.Window.ScreenshotManager).ToString())
				{
                    return Runtime.Windows.screenshotPanel;
				}
			}
			catch (Exception ex)
			{
                Runtime.Log.Error("GetContentFromPersistString failed" + Constants.vbNewLine + ex.Message);
			}
					
			return null;
		}
        #endregion
	}
}
